cpp/src/slice2cs/Gen.cpp

```diff
@@ -157,110 +157,191 @@ Slice::CsVisitor::~CsVisitor()
+                             member->type(),
+                             bitSequenceIndex,
+                             false,
+                             ns,
+                             members.size() == 1 ? "value" : "value." + fieldName(member),
+                             stream);
+        }
+        if (bitSequenceSize > 0)
+        {
+            assert(static_cast<size_t>(bitSequenceIndex) == bitSequenceSize);
+        }
 
-    for (const auto& param : requiredParams)
-    {
-        writeMarshalCode(
-            _out, param->type(), bitSequenceIndex, false, ns, singleParam ? "value" : obj + fieldName(param), stream);
-    }
-    if (bitSequenceSize > 0)
-    {
-        assert(static_cast<size_t>(bitSequenceIndex) == bitSequenceSize);
+        for (const auto& member : taggedMembers)
+        {
+            writeTaggedMarshalCode(_out,
+                                   OptionalPtr::dynamicCast(member->type()),
+                                   false,
+                                   ns,
+                                   members.size() == 1 ? "value" : "value." + fieldName(member), member->tag(),
+                                   stream);
+        }
+
+        if (repeat)
+        {
+            repeat = false;
+        }
+        else if (write11ReturnLast)
+        {
+            _out << eb;
+            _out << nl;
+            _out << "else";
+            _out << sb;
+
+            // Repeat after updating requiredMembers
+            repeat = true;
+            requiredMembers.push_back(requiredMembers.front());
+            requiredMembers.pop_front();
+        }
     }
+    while (repeat)
+        ;
 
-    for (const auto& param : taggedParams)
+    if (write11ReturnLast)
     {
-        writeTaggedMarshalCode(_out,
-                               OptionalPtr::dynamicCast(param->type()),
-                               false,
-                               ns,
-                               singleParam ? "value" : obj + fieldName(param), param->tag(),
-                               stream);
+        _out << eb;
     }
 }
 
 void
-Slice::CsVisitor::writeUnmarshalParams(const OperationPtr& op,
-                                       const MemberList& requiredParams,
-                                       const MemberList& taggedParams,
-                                       bool isReturnValue,
-                                       const string& customStream)
+Slice::CsVisitor::writeUnmarshal(const OperationPtr& operation, bool returnType)
 {
-    const string stream = customStream.empty() ? "istr" : customStream;
-    string ns = getNamespace(op->interface());
+    const string stream = "istr";
+    string ns = getNamespace(operation->interface());
+
+    MemberList members = returnType ? operation->returnValues() : operation->parameters();
+    auto [requiredMembers, taggedMembers] = getSortedMembers(members);
 
     int bitSequenceIndex = -1;
     size_t bitSequenceSize = 0;
```

Comment:
This could also be reduced to a single size.

Reply:
Fixed.